File: orcastork/adapters/redis/gate.py
"""Redis ``CooldownGate`` — ``SET NX PX``: the key exists exactly while the cooldown runs.

Redis owns the expiry, so the cooldown holds across pods and restarts by construction; the
``NX`` set is the atomic check-and-arm.
"""

from __future__ import annotations

from typing import Any

from redis.asyncio import Redis


class RedisCooldownGate:
    def __init__(self, redis: Redis) -> None:
        self._redis: Any = redis  # see RedisDataPointStore: redis-py's sync/async union typing

    async def try_acquire(self, key: str, cooldown_seconds: float) -> bool:
        return bool(await self._redis.set(f'cooldown:{key}', '1', nx=True, px=max(int(cooldown_seconds * 1000), 1)))

File: typescript/src/orcastork/adapters/redis/gate.ts
/**
 * Redis `CooldownGate` — `SET NX PX`: the key exists exactly while the cooldown runs.
 *
 * Redis owns the expiry, so the cooldown holds across pods and restarts by construction; the `NX`
 * set is the atomic check-and-arm.
 *
 * @module
 */

import type { CooldownGate } from '../../ports/cooldown_gate.js';

/** The one command this gate needs. */
export interface RedisGateClient {
  /** `SET key value [NX] [PX ms]` — the reply is `'OK'` when the set happened, `null` when `NX` lost. */
  set(
    key: string,
    value: string,
    options: { condition: 'NX'; expiration: { type: 'PX'; value: number } },
  ): Promise<unknown>;
}

/** Check-and-arm in one step, with the armed windows owned by Redis. */
export class RedisCooldownGate implements CooldownGate {
  private readonly redis: RedisGateClient;

  public constructor(redis: RedisGateClient) {
    this.redis = redis;
  }

  public async tryAcquire(key: string, cooldownMs: number): Promise<boolean> {
    // At least 1ms: `PX 0` is rejected by Redis, so a zero/sub-millisecond cooldown still arms a
    // key that expires immediately rather than failing the call.
    const expiration = Math.max(Math.trunc(cooldownMs), 1);
    const reply = await this.redis.set(`cooldown:${key}`, '1', {
      condition: 'NX',
      expiration: { type: 'PX', value: expiration },
    });
    return reply !== null && reply !== undefined;
  }
}
